<a href="https://colab.research.google.com/github/schuylerdare96/ML-AI-Data-Science-Projects/blob/master/bulldozer_price_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [ ]:
df = pd.read_csv("/content/drive/My Drive/Data for Colab notebooks/Bulldozer price prediction dataset/TrainAndValid.csv",
                 low_memory=False,
                 parse_dates=["saledate"])

In [ ]:
df.head()

SalesID  SalePrice  ...  Differential_Type  Steering_Controls
0  1139246    66000.0  ...           Standard       Conventional
1  1139248    57000.0  ...           Standard       Conventional
2  1139249    10000.0  ...                NaN                NaN
3  1139251    38500.0  ...                NaN                NaN
4  1139253    11000.0  ...                NaN                NaN

[5 rows x 53 columns]

In [ ]:
df.sort_values(by=["saledate"], inplace=True, ascending=True)
df.saledate.head(20)

205615   1989-01-17
274835   1989-01-31
141296   1989-01-31
212552   1989-01-31
62755    1989-01-31
54653    1989-01-31
81383    1989-01-31
204924   1989-01-31
135376   1989-01-31
113390   1989-01-31
113394   1989-01-31
116419   1989-01-31
32138    1989-01-31
127610   1989-01-31
76171    1989-01-31
127000   1989-01-31
128130   1989-01-31
127626   1989-01-31
55455    1989-01-31
55454    1989-01-31
Name: saledate, dtype: datetime64[ns]

In [ ]:
df_tmp = df.copy()
df_tmp.head()

SalesID  SalePrice  ...  Differential_Type  Steering_Controls
205615  1646770     9500.0  ...                NaN                NaN
274835  1821514    14000.0  ...           Standard       Conventional
141296  1505138    50000.0  ...                NaN                NaN
212552  1671174    16000.0  ...           Standard       Conventional
62755   1329056    22000.0  ...                NaN                NaN

[5 rows x 53 columns]

In [ ]:
df_tmp["saleYear"] = df_tmp.saledate.dt.year
df_tmp["saleMonth"] = df_tmp.saledate.dt.month
df_tmp["saleDay"] = df_tmp.saledate.dt.day
df_tmp["saleDayofweek"] = df_tmp.saledate.dt.dayofweek
df_tmp["saleDayofyear"] = df_tmp.saledate.dt.dayofyear

# Drop original saledate
df_tmp.drop("saledate", axis=1, inplace=True)

In [ ]:
df_tmp.head().T

205615  ...                                             62755 
SalesID                                                             1646770  ...                                            1329056
SalePrice                                                              9500  ...                                              22000
MachineID                                                           1126363  ...                                            1336053
ModelID                                                                8434  ...                                               4089
datasource                                                              132  ...                                                132
auctioneerID                                                             18  ...                                                 99
YearMade                                                               1974  ...                                               1984
MachineHoursCurrentMeter                                                NaN  ...                                                NaN
UsageBand                                                               NaN  ...                                                NaN
fiModelDesc                                                            TD20  ...                                                D3B
fiBaseModel                                                            TD20  ...                                                 D3
fiSecondaryDesc                                                         NaN  ...                                                  B
fiModelSeries                                                           NaN  ...                                                NaN
fiModelDescriptor                                                       NaN  ...                                                NaN
ProductSize                                                          Medium  ...                                                NaN
fiProductClassDesc        Track Type Tractor, Dozer - 105.0 to 130.0 Hor...  ...  Track Type Tractor, Dozer - 20.0 to 75.0 Horse...
state                                                                 Texas  ...                                            Florida
ProductGroup                                                            TTT  ...                                                TTT
ProductGroupDesc                                        Track Type Tractors  ...                                Track Type Tractors
Drive_System                                                            NaN  ...                                                NaN
Enclosure                                                             OROPS  ...                                              OROPS
Forks                                                                   NaN  ...                                                NaN
Pad_Type                                                                NaN  ...                                                NaN
Ride_Control                                                            NaN  ...                                                NaN
Stick                                                                   NaN  ...                                                NaN
Transmission                                                   Direct Drive  ...                                           Standard
Turbocharged                                                            NaN  ...                                                NaN
Blade_Extension                                                         NaN  ...                                                NaN
Blade_Width                                                             NaN  ...                                                NaN
Enclosure_Type                                                          NaN  ...                             

In [ ]:
df.tail().T

410879  ...                                    409203
SalesID                                                             6302984  ...                                   6283635
SalePrice                                                             16000  ...                                     34000
MachineID                                                           1915521  ...                                   1869284
ModelID                                                                5266  ...                                      4701
datasource                                                              149  ...                                       149
auctioneerID                                                             99  ...                                        99
YearMade                                                               2001  ...                                      1000
MachineHoursCurrentMeter                                                NaN  ...                                       NaN
UsageBand                                                               NaN  ...                                       NaN
saledate                                                2012-04-28 00:00:00  ...                       2012-04-28 00:00:00
fiModelDesc                                                            D38E  ...                                      544H
fiBaseModel                                                             D38  ...                                       544
fiSecondaryDesc                                                           E  ...                                         H
fiModelSeries                                                           NaN  ...                                       NaN
fiModelDescriptor                                                       NaN  ...                                       NaN
ProductSize                                                             NaN  ...                                       NaN
fiProductClassDesc        Track Type Tractor, Dozer - 75.0 to 85.0 Horse...  ...  Wheel Loader - 120.0 to 135.0 Horsepower
state                                                            California  ...                                California
ProductGroup                                                            TTT  ...                                        WL
ProductGroupDesc                                        Track Type Tractors  ...                              Wheel Loader
Drive_System                                                            NaN  ...                                       NaN
Enclosure                                                             OROPS  ...                                EROPS w AC
Forks                                                                   NaN  ...                       None or Unspecified
Pad_Type                                                                NaN  ...                                       NaN
Ride_Control                                                            NaN  ...                       None or Unspecified
Stick                                                                   NaN  ...                                       NaN
Transmission                                                       Standard  ...                                       NaN
Turbocharged                                                            NaN  ...                                       NaN
Blade_Extension                                                         NaN  ...                                       NaN
Blade_Width                                                             NaN  ...                                       NaN
Enclosure_Type                                                          NaN  ...                                       NaN
Engine_Horsepower                                                       NaN  ...                                       NaN
Hydraulics

In [ ]:
df_tmp.ProductGroupDesc.value_counts()

Track Excavators       104230
Track Type Tractors     82582
Backhoe Loaders         81401
Wheel Loader            73216
Skid Steer Loaders      45011
Motor Graders           26258
Name: ProductGroupDesc, dtype: int64

In [ ]:
df_tmp.state.value_counts()

Florida           67320
Texas             53110
California        29761
Washington        16222
Georgia           14633
Maryland          13322
Mississippi       13240
Ohio              12369
Illinois          11540
Colorado          11529
New Jersey        11156
North Carolina    10636
Tennessee         10298
Alabama           10292
Pennsylvania      10234
South Carolina     9951
Arizona            9364
New York           8639
Connecticut        8276
Minnesota          7885
Missouri           7178
Nevada             6932
Louisiana          6627
Kentucky           5351
Maine              5096
Indiana            4124
Arkansas           3933
New Mexico         3631
Utah               3046
Unspecified        2801
Wisconsin          2745
New Hampshire      2738
Virginia           2353
Idaho              2025
Oregon             1911
Michigan           1831
Wyoming            1672
Iowa               1336
Montana            1336
Oklahoma           1326
Nebraska            866
West Virginia   

In [ ]:
np.random.seed(69)

from sklearn.ensemble import RandomForestRegressor

model=RandomForestRegressor(n_jobs=-1) #n_jobs decides how many cpu cores are 2b used; -1=all cpu cores 2b used.

## the model needs 2b converted into numbers; 

* this is done using pandas category-a datatype of pandas

In [ ]:
for key,value in df_tmp.items():
    if pd.api.types.is_string_dtype(value):
        df_tmp[key]=value.astype("category").cat.as_ordered()
        
# cat is used to represent categories in oanda just as date is represented by dt 
# the above function treats each label as key & its content as value, traverses through the dictionary, converts each value in2 string where dtypr is number       

In [ ]:
df_tmp.dtypes

SalesID                        int64
SalePrice                    float64
MachineID                      int64
ModelID                        int64
datasource                     int64
auctioneerID                 float64
YearMade                       int64
MachineHoursCurrentMeter     float64
UsageBand                   category
fiModelDesc                 category
fiBaseModel                 category
fiSecondaryDesc             category
fiModelSeries               category
fiModelDescriptor           category
ProductSize                 category
fiProductClassDesc          category
state                       category
ProductGroup                category
ProductGroupDesc            category
Drive_System                category
Enclosure                   category
Forks                       category
Pad_Type                    category
Ride_Control                category
Stick                       category
Transmission                category
Turbocharged                category
B

In [ ]:
df_tmp.fiModelDesc.cat.codes.head(20)

205615    4592
274835    1819
141296    2347
212552    1818
62755     2118
54653     2297
81383     3125
204924     978
135376    2223
113390      81
113394      81
116419     321
32138     1062
127610    1738
76171     1782
127000    1696
128130    1738
127626    1738
55455       81
55454       81
dtype: int16

In [ ]:
df_tmp.isna().sum()

SalesID                          0
SalePrice                        0
MachineID                        0
ModelID                          0
datasource                       0
auctioneerID                 20136
YearMade                         0
MachineHoursCurrentMeter    265194
UsageBand                   339028
fiModelDesc                      0
fiBaseModel                      0
fiSecondaryDesc             140727
fiModelSeries               354031
fiModelDescriptor           337882
ProductSize                 216605
fiProductClassDesc               0
state                            0
ProductGroup                     0
ProductGroupDesc                 0
Drive_System                305611
Enclosure                      334
Forks                       214983
Pad_Type                    331602
Ride_Control                259970
Stick                       331602
Transmission                224691
Turbocharged                331602
Blade_Extension             386715
Blade_Width         

In [ ]:
df_tmp.to_csv("/content/drive/My Drive/Data for Colab notebooks/Bulldozer price prediction dataset/modified_df.csv",index=False)

In [ ]:

df_tmp=pd.read_csv("/content/drive/My Drive/Data for Colab notebooks/Bulldozer price prediction dataset/modified_df.csv",low_memory=False)

In [ ]:
df_tmp.dtypes

SalesID                       int64
SalePrice                   float64
MachineID                     int64
ModelID                       int64
datasource                    int64
auctioneerID                float64
YearMade                      int64
MachineHoursCurrentMeter    float64
UsageBand                    object
fiModelDesc                  object
fiBaseModel                  object
fiSecondaryDesc              object
fiModelSeries                object
fiModelDescriptor            object
ProductSize                  object
fiProductClassDesc           object
state                        object
ProductGroup                 object
ProductGroupDesc             object
Drive_System                 object
Enclosure                    object
Forks                        object
Pad_Type                     object
Ride_Control                 object
Stick                        object
Transmission                 object
Turbocharged                 object
Blade_Extension             

In [ ]:
for key,value in df_tmp.items():
    if pd.api.types.is_string_dtype(value):
        df_tmp[key]=value.astype("category").cat.as_ordered()

In [ ]:
df_tmp.dtypes

SalesID                        int64
SalePrice                    float64
MachineID                      int64
ModelID                        int64
datasource                     int64
auctioneerID                 float64
YearMade                       int64
MachineHoursCurrentMeter     float64
UsageBand                   category
fiModelDesc                 category
fiBaseModel                 category
fiSecondaryDesc             category
fiModelSeries               category
fiModelDescriptor           category
ProductSize                 category
fiProductClassDesc          category
state                       category
ProductGroup                category
ProductGroupDesc            category
Drive_System                category
Enclosure                   category
Forks                       category
Pad_Type                    category
Ride_Control                category
Stick                       category
Transmission                category
Turbocharged                category
B

In [ ]:
df_tmp.ProductGroupDesc.cat.codes

0         4
1         5
2         4
3         5
4         4
         ..
412693    4
412694    2
412695    3
412696    1
412697    5
Length: 412698, dtype: int8

In [ ]:
for key, values in df_tmp.items():
    if pd.api.types.is_numeric_dtype(values):
        print(key)

SalesID
SalePrice
MachineID
ModelID
datasource
auctioneerID
YearMade
MachineHoursCurrentMeter
saleYear
saleMonth
saleDay
saleDayofweek
saleDayofyear


In [ ]:
for key,values in df_tmp.items():
    if pd.api.types.is_numeric_dtype(values):
        if pd.isnull(values).sum():
            print(key)

auctioneerID
MachineHoursCurrentMeter


In [ ]:
for key,values in df_tmp.items():
    if pd.api.types.is_numeric_dtype(values):
        if pd.isnull(values).sum():
            df_tmp[key+"_is_missing"]=pd.isnull(values)
            df_tmp[key]=values.fillna(values.median())

In [ ]:
for key,values in df_tmp.items():
    if pd.api.types.is_numeric_dtype(values):
        if pd.isnull(values).sum():
            print(key)

In [ ]:
df_tmp.head()


SalesID  ...  MachineHoursCurrentMeter_is_missing
0  1646770  ...                                 True
1  1821514  ...                                 True
2  1505138  ...                                 True
3  1671174  ...                                 True
4  1329056  ...                                 True

[5 rows x 59 columns]

In [ ]:
df_tmp.auctioneerID_is_missing.value_counts(
)

False    392562
True      20136
Name: auctioneerID_is_missing, dtype: int64

In [ ]:
df_tmp.isna().sum()

SalesID                                     0
SalePrice                                   0
MachineID                                   0
ModelID                                     0
datasource                                  0
auctioneerID                                0
YearMade                                    0
MachineHoursCurrentMeter                    0
UsageBand                              339028
fiModelDesc                                 0
fiBaseModel                                 0
fiSecondaryDesc                        140727
fiModelSeries                          354031
fiModelDescriptor                      337882
ProductSize                            216605
fiProductClassDesc                          0
state                                       0
ProductGroup                                0
ProductGroupDesc                            0
Drive_System                           305611
Enclosure                                 334
Forks                             

In [ ]:
#Lists out all categorical datatype in the DataFrame
pd.Categorical(df_tmp["state"]).codes

array([43,  8,  8, ...,  4,  4,  4], dtype=int8)

In [ ]:
for key,value in df_tmp.items():
    if not pd.api.types.is_numeric_dtype(value):
        df_tmp[key+"_is_missing"]=pd.isnull(value)
        df_tmp[key]=pd.Categorical(value).codes+1
        
# +1 is done because if a categorical value is null, pandas assigns -1; thus +1 converts it to 0.

In [ ]:
df_tmp.head()

SalesID  ...  Steering_Controls_is_missing
0  1646770  ...                          True
1  1821514  ...                         False
2  1505138  ...                          True
3  1671174  ...                         False
4  1329056  ...                          True

[5 rows x 103 columns]

In [ ]:
df_tmp.head().T

0        1        2        3        4
SalesID                       1646770  1821514  1505138  1671174  1329056
SalePrice                        9500    14000    50000    16000    22000
MachineID                     1126363  1194089  1473654  1327630  1336053
ModelID                          8434    10150     4139     8591     4089
datasource                        132      132      132      132      132
...                               ...      ...      ...      ...      ...
Backhoe_Mounting_is_missing     False     True    False     True    False
Blade_Type_is_missing           False     True    False     True    False
Travel_Controls_is_missing      False     True    False     True    False
Differential_Type_is_missing     True    False     True    False     True
Steering_Controls_is_missing     True    False     True    False     True

[103 rows x 5 columns]

In [ ]:
df_tmp.isna().sum()[:103]

SalesID                         0
SalePrice                       0
MachineID                       0
ModelID                         0
datasource                      0
                               ..
Backhoe_Mounting_is_missing     0
Blade_Type_is_missing           0
Travel_Controls_is_missing      0
Differential_Type_is_missing    0
Steering_Controls_is_missing    0
Length: 103, dtype: int64

In [ ]:
df_tmp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 412698 entries, 0 to 412697
Columns: 103 entries, SalesID to Steering_Controls_is_missing
dtypes: bool(46), float64(3), int16(4), int64(10), int8(40)
memory usage: 77.9 MB


# modelling 

In [ ]:
%%time
from sklearn.ensemble import RandomForestRegressor

np.random.seed(69)

model=RandomForestRegressor(n_jobs=-1)

model.fit(df_tmp.drop("SalePrice",axis=1),df_tmp["SalePrice"])


CPU times: user 19min 48s, sys: 3.92 s, total: 19min 52s
Wall time: 10min 8s


In [ ]:
model.score(df_tmp.drop("SalePrice",axis=1),df_tmp["SalePrice"])

0.9876095059066801

In [ ]:
df_tmp.isna().sum().head()

SalesID       0
SalePrice     0
MachineID     0
ModelID       0
datasource    0
dtype: int64

In [ ]:
df_tmp.saleYear.value_counts()

2009    43849
2008    39767
2011    35197
2010    33390
2007    32208
2006    21685
2005    20463
2004    19879
2001    17594
2000    17415
2002    17246
2003    15254
1998    13046
1999    12793
2012    11573
1997     9785
1996     8829
1995     8530
1994     7929
1993     6303
1992     5519
1991     5109
1989     4806
1990     4529
Name: saleYear, dtype: int64

In [ ]:
df_val=df_tmp[df_tmp.saleYear==2012]
df_train=df_tmp[df_tmp.saleYear !=2012]

In [ ]:
X_train,y_train=df_train.drop("SalePrice",axis=1),df_train.SalePrice
X_val,y_val=df_val.drop("SalePrice",axis=1),df_val.SalePrice

In [ ]:
X_train.shape,y_train.shape,X_val.shape,y_val.shape

((401125, 102), (401125,), (11573, 102), (11573,))

In [ ]:
from sklearn.metrics import mean_squared_log_error, mean_absolute_error,r2_score

def rmsle(y_test,y_preds):
    """
    This function finds out Root Mean Square Log Error
    """
    return np.sqrt(mean_squared_log_error(y_test,y_preds))
    
def show_scores(model):
    
    train_preds=model.predict(X_train)
    val_preds=model.predict(X_val)
    
    scores={"Training MAE":mean_absolute_error(y_train,train_preds),
           "Valid MAE":mean_absolute_error(y_val,val_preds),
           "Training RMSLE":rmsle(y_train,train_preds),
           "Valid RMSLE":rmsle(y_val,val_preds),
           "Training R^2 score":r2_score(y_train,train_preds),
           "Valid R^2":r2_score(y_val,val_preds)}
    return scores

    

In [ ]:
df_tmp.dtypes

SalesID                           int64
SalePrice                       float64
MachineID                         int64
ModelID                           int64
datasource                        int64
                                 ...   
Backhoe_Mounting_is_missing        bool
Blade_Type_is_missing              bool
Travel_Controls_is_missing         bool
Differential_Type_is_missing       bool
Steering_Controls_is_missing       bool
Length: 103, dtype: object

In [ ]:
X_train.shape[1]

102

In [ ]:
%%time
model=RandomForestRegressor(n_jobs=-1,max_samples=10000)

# max_samples decides on how many tables will our model be tested
# n_estimators= no of estimators our ml model will be tried on.

model.fit(X_train,y_train)

CPU times: user 43.5 s, sys: 285 ms, total: 43.8 s
Wall time: 23.2 s


In [ ]:
show_scores(model)

{'Training MAE': 5559.580766344656,
 'Training RMSLE': 0.25744850010501225,
 'Training R^2 score': 0.8607038553911985,
 'Valid MAE': 7130.927808692647,
 'Valid RMSLE': 0.29154771960428155,
 'Valid R^2': 0.8345679113406443}

In [ ]:
%%time

from sklearn.model_selection import RandomizedSearchCV
np.random.seed(69)

rf_grid={
    "n_estimators":np.arange(10,100,10),
    "max_depth":[None,3,5,10],
    "min_samples_split":np.arange(2,20,2),
    "min_samples_leaf":np.arange(1,20,2),
    "max_features":[0.5,1,"sqrt","auto"],
    "max_samples":[10000]
}


rs_model = RandomizedSearchCV(RandomForestRegressor(n_jobs=-1),
                                                    param_distributions=rf_grid,
                                                    n_iter=2,
                                                    cv=5,
                                                    verbose=True)

rs_model.fit(X_train,y_train)

Fitting 5 folds for each of 2 candidates, totalling 10 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:   25.1s finished


CPU times: user 21.2 s, sys: 1.62 s, total: 22.8 s
Wall time: 27.3 s


In [ ]:
rs_model.best_params_

{'max_depth': 3,
 'max_features': 0.5,
 'max_samples': 10000,
 'min_samples_leaf': 19,
 'min_samples_split': 18,
 'n_estimators': 10}

In [ ]:
%%time
np.random.seed(69)

ideal_model=RandomForestRegressor(n_estimators=40,
                                 min_samples_leaf=1,
                                 min_samples_split=14,
                                 max_features=0.5,
                                 n_jobs=-1,
                                 max_samples=None,
                                 )
ideal_model.fit(X_train,y_train)


CPU times: user 3min 29s, sys: 254 ms, total: 3min 29s
Wall time: 1min 46s


In [ ]:
%%time

show_scores(ideal_model)

CPU times: user 8.15 s, sys: 124 ms, total: 8.28 s
Wall time: 7.03 s


{'Training MAE': 2948.430573252706,
 'Training RMSLE': 0.14432392169454122,
 'Training R^2 score': 0.9590579495764435,
 'Valid MAE': 5940.492994228834,
 'Valid RMSLE': 0.24591864562715962,
 'Valid R^2': 0.8823697569318704}

In [ ]:
df_test=pd.read_csv("/content/drive/My Drive/Data for Colab notebooks/Bulldozer price prediction dataset/Test.csv",low_memory=False,parse_dates=["saledate"])

df_test.head()
df_test.dtypes



SalesID                              int64
MachineID                            int64
ModelID                              int64
datasource                           int64
auctioneerID                         int64
YearMade                             int64
MachineHoursCurrentMeter           float64
UsageBand                           object
saledate                    datetime64[ns]
fiModelDesc                         object
fiBaseModel                         object
fiSecondaryDesc                     object
fiModelSeries                       object
fiModelDescriptor                   object
ProductSize                         object
fiProductClassDesc                  object
state                               object
ProductGroup                        object
ProductGroupDesc                    object
Drive_System                        object
Enclosure                           object
Forks                               object
Pad_Type                            object
Ride_Contro

## The data needs to be preprocessed before testing, since the demo data has 102 columns & test data has 52 columns.

In [ ]:
%%time
df_test1=pd.read_csv("/content/drive/My Drive/Data for Colab notebooks/Bulldozer price prediction dataset/Test.csv",low_memory=False,parse_dates=["saledate"])

def preprocess_data(df_test):
    
    """
    This function converts non-integral data into categorical data
    This also fills empty integer columns with median
    """
    df_test["saleYear"] = df_test.saledate.dt.year
    df_test["saleMonth"] = df_test.saledate.dt.month
    df_test["saleDay"] = df_test.saledate.dt.day
    df_test["saleDayofweek"] = df_test.saledate.dt.dayofweek
    df_test["saleDayofyear"] = df_test.saledate.dt.dayofyear

    # Drop original saledate
    df_test.drop("saledate", axis=1, inplace=True)
    
    for key,values in df_test.items():
        if pd.api.types.is_numeric_dtype(values):
            if pd.isnull(values).sum():
                df_test[key+"_is_missing"]=pd.isnull(values)
                df_test[key]=values.fillna(values.median())
            
        if not pd.api.types.is_numeric_dtype(values):
                df_test[key+"_is_missing"]=pd.isnull(values)
                df_test[key]=pd.Categorical(values).codes+1
            
    return df_test

CPU times: user 90.1 ms, sys: 3 ms, total: 93.1 ms
Wall time: 98.4 ms


In [ ]:
df_finalTest=preprocess_data(df_test1)

In [ ]:
df_test1.head().T

0        1        2        3        4
SalesID                       1227829  1227844  1227847  1227848  1227863
MachineID                     1006309  1022817  1031560    56204  1053887
ModelID                          3168     7271    22805     1269    22312
datasource                        121      121      121      121      121
auctioneerID                        3        3        3        3        3
...                               ...      ...      ...      ...      ...
Backhoe_Mounting_is_missing      True     True     True     True    False
Blade_Type_is_missing            True     True     True     True    False
Travel_Controls_is_missing       True     True     True     True    False
Differential_Type_is_missing     True    False     True     True     True
Steering_Controls_is_missing     True    False     True     True     True

[101 rows x 5 columns]

In [ ]:
ideal_model.predict(df_test1)

ValueError: ignored

In [ ]:
# know hwich columns are missing from original trained data from test data.

set(X_train.columns)-set(df_test1.columns)

In [ ]:
df_test1.isna().sum()

In [ ]:
df_test1["auctioneerID_is_missing"]=False

In [ ]:
df_test1

In [ ]:
df_test1.isna().sum()

In [ ]:
%%time
test_preds=ideal_model.predict(df_test1)
test_preds

In [ ]:
df_final=pd.DataFrame()
df_final["SalesID"]=df_test1["SalesID"]
df_final["SalesPrice"]=test_preds
df_final.T

In [ ]:
df_final.to_csv("data/bluebook-for-bulldozers/df_final_results")

In [ ]:
ideal_model.feature_importances_

In [ ]:
df_test1.dtypes.T

In [ ]:
import seaborn as sns

def plot_features(columns,importances,n=20):
    df=(pd.DataFrame({"features":columns,
                     "feature_importances":importances})
                .sort_values("feature_importances",ascending=False)
                   .reset_index(drop=True))
    plt.figure(figsize=(20,10))
    fig,ax=plt.subplots()
    ax.barh(df["features"][:n],df["feature_importances"][:20])
    ax.set_ylabel("features")
    ax.set_xlabel("feature importances")
    
    sns.barplot(x="feature_importances",
                y="features",
                data=df[:n],
                orient="h")
    

In [ ]:
plot_features(X_train.columns,ideal_model.feature_importances_)